main.go

```diff
@@ -190,13 +252,106 @@ Examples:
 
 See the repo for more details: https://github.com/multiprocessio/dsq.`
 
+func importFilesIteration(cacheMode bool, filesPtr *[]string, ec *runner.EvalContext, project *runner.ProjectState, i int) (int, error) {
+	files := *filesPtr
+	file := files[i]
+	panelId := uuid.New().String()
+	resultFile := ec.GetPanelResultsFile(project.Id, panelId)
+	out, err := openTruncate(resultFile)
+	if err != nil {
+		return i, err
+	}
+	defer out.Close()
+
+	if file == "" {
+		b, err := ioutil.ReadAll(os.Stdin)
+		if err == nil {
+			if i == len(files)-1 {
+				return i, errors.New("Expected file extension or mimetype: e.g. cat x.csv | dsq -s csv, or cat x.csv | dsq -s text/csv")
+			}
+			mimetype := files[i+1]
+			if !strings.Contains(mimetype, string(filepath.Separator)) {
+				mimetype = string(runner.GetMimeType("x."+mimetype, runner.ContentTypeInfo{}))
+			}
+
+			if mimetype == "" {
+				return i, fmt.Errorf("Unknown mimetype or file extension: %s.", mimetype)
+			}
+			i += 1
+
+			var r io.Reader = bytes.NewReader(b)
+			if cacheMode {
+				r, err = os.Open(files[len(files)-1])
+				if err != nil {
+					return i, fmt.Errorf("Error opening copied stdin file: %v", err) // returns as it's not possible to go back to default.
+				} else {
+					*filesPtr = files[:len(files)-1]
+				}
+			}
+			fmt.Println(files)
+
+			cti := runner.ContentTypeInfo{Type: string(mimetype)}
+			err := runner.TransformReader(r, "", cti, out)
+			if err != nil {
+				return i, err
+			}
+
+			s, err := getShape(resultFile, "")
+			if err != nil {
+				return i, err
+			}
+
+			project.Pages[0].Panels = append(project.Pages[0].Panels, runner.PanelInfo{
+				ResultMeta: runner.PanelResult{
+					Shape: *s,
+				},
+				Id:   panelId,
+				Name: file,
+			})
+
+		}
+
+		return i, nil
+	}
+
+	if err := evalFileInto(file, out); err != nil {
+		return i, err
+	}
+
+	s, err := getShape(resultFile, file)
+	if err != nil {
+		return i, err
+	}
+
+	project.Pages[0].Panels = append(project.Pages[0].Panels, runner.PanelInfo{
+		ResultMeta: runner.PanelResult{
+			Shape: *s,
+		},
+		Id:   panelId,
+		Name: uuid.New().String(),
+	})
+
+	return i, nil
+}
+
+func importFiles(cacheMode bool, files []string, ec *runner.EvalContext, project *runner.ProjectState) error {
+	var err error
+	for i := 0; i < len(files); i++ {
+		if i, err = importFilesIteration(cacheMode, &files, ec, project, i); err != nil {
```

Comment:
Why do you need to pass in files and i separately? Why can't you just pass in `files[i]`?

Also, you can just call this `importFile` rather than `importFilesIteration` unless there's a reason not to use `importFile`.

Reply:
Yeah I see how you depend on `i` inside this function but I don't think that's a good idea.

I'm really sorry about dragging out this PR though we can merge it and I can do some followup cleanup.

I don't want you to keep having to edit stuff.

Reply:
We need to modify the files slice (delete the tmp file for stdin copying) in cache mode.

Reply:
And yeah it could be importFile

Reply:
Yeah if you don't mind, you can merge it.

Reply:
Ok! I'll open a PR for it after this is merged so you can comment on it if you want.

I'm ok to merge this now then.

Reply:
Ready to merge the DataStation PR first?

Reply:
Yeah!

Reply:
Ok, that's merged. Update go.mod in here and then we'll merge this

Reply:
Done.